### PR TITLE
HIR dialect

### PR DESCRIPTION
**Pull request by kingshukmajumder**
PR for merging HIR dialect to CIRCT main branch.

### CODE REVIEW

File: .github/workflows/buildAndTest.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ name: Build and Test
 on:
   push:
     branches:
-      - main
+      - hir-dev
```
Comment (lattner) [inline .github/workflows/buildAndTest.yml:6]: Please don't switch branches for tests here.

File: .github/workflows/buildAndTest.yml [hunk 2]
```diff
@@ -363,4 +363,5 @@ jobs:
             echo "================"
           fi
 
+>>>>>>> 11431bf7d21171850933d2ff546e28fd4720613b
```
Comment (lattner) [inline .github/workflows/buildAndTest.yml:366]: merge conflict?

File: lib/Dialect/HIR/IR/HIROpInterfaces.cpp [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#include "circt/Dialect/HIR/IR/HIR.h"
```
Comment (lattner) [inline lib/Dialect/HIR/IR/HIROpInterfaces.cpp:1]: Please add header comment boilerplate per the dev policy

File: lib/Dialect/HIR/IR/HIROpSyntax.cpp [hunk 4]
```diff
@@ -0,0 +1,333 @@
+#include "circt/Dialect/HIR/IR/HIROpSyntax.h"
+#include "circt/Dialect/HIR/IR/helper.h"
+#define min(x, y) x > y ? x : y
```
Comment (lattner) [inline lib/Dialect/HIR/IR/HIROpSyntax.cpp:3]: use std::min ?

File: lib/Dialect/HIR/IR/HIRTypeSyntax.cpp [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+    if (parser.parseRParen())
+      return Type();
+  }
+
+  // Process output types and delays.
+  SmallVector<Type, 4> resultTypes;
+  if (!parser.parseOptionalArrow()) {
+    if (parser.parseLParen())
+      return Type();
+    if (parser.parseOptionalRParen()) {
```
Comment (lattner) [inline lib/Dialect/HIR/IR/HIRTypeSyntax.cpp:188]: Can this use the comma separated list parser hooks?

**Review by lattner — COMMENTED** [review]
Just some quick comments, I didn't do a detailed review

**Comment by mikeurbach** [comment]
First, I have some logistical suggestions, since this encompasses a few things. I don't know what the policy is for wholesale additions of a dialect, but this is quite a bit to review at once. It might help to separate this into several patches along these lines:

* Boilerplate to create new dialect without IR, Analysis, Transforms, Conversions, or Translations
* IR with roundtrip lit tests through circt-opt
* The rest of the libraries, in separate PRs, with their respective tests

Regardless, I will try to take a deeper look at this. For now I had a couple high-level questions.

Could the `ScheduleAnalysis` make use of CIRCT's [infrastructure](https://github.com/llvm/circt/blob/main/include/circt/Scheduling/Problems.h) for modeling scheduling problems and algorithms? I think it would be great to (eventually) develop what is needed for HIR in the common infrastructure if it doesn't exist yet.

I saw a mention of `createStandardToHIRPass` but I didn't see its implementation. In general, I am curious what dialects you are interested in lowering to HIR. I am currently interested in using the scheduling infrastructure on affine loops, and then lowering these to this [pipeline while op](https://circt.llvm.org/docs/Dialects/StaticLogic/#staticlogicpipelinewhile-circtstaticlogicpipelinewhileop) representation, which captures dataflow but leaves exact timing to be determined by lowering to Calyx (which is in progress). Another design point we have considered on this path is using something like HIR, which I am interested in exploring. Once the dialect lands, I'd be interested in doing something like the above but targetting HIR instead of the pipeline while op.

File: .gitignore [hunk 6]
```diff
@@ -18,3 +18,12 @@ lit.site.cfg.py
...
 /ext
+
+#Misc
+compile_commands.json
+.clangd
+./*.sv
+*.log
+*.pb
+xsim.dir
+.cache
```
Comment (darthscsi) [inline .gitignore:29]: Changes seem unrelated to adding a dialect.

File: CMakeLists.txt [hunk 7]
```diff
@@ -6,6 +6,8 @@
 
 cmake_minimum_required(VERSION 3.13.4)
 
+set( CMAKE_EXPORT_COMPILE_COMMANDS ON )
```
Comment (darthscsi) [inline CMakeLists.txt:9]: This already exists in the cmake file.

**Comment by kingshukmajumder** [comment]
> First, I have some logistical suggestions, since this encompasses a few things. I don't know what the policy is for wholesale additions of a dialect, but this is quite a bit to review at once. It might help to separate this into several patches along these lines:
> 
> * Boilerplate to create new dialect without IR, Analysis, Transforms, Conversions, or Translations
> * IR with roundtrip lit tests through circt-opt
> * The rest of the libraries, in separate PRs, with their respective tests
> 
> Regardless, I will try to take a deeper look at this. For now I had a couple high-level questions.
> 
> Could the `ScheduleAnalysis` make use of CIRCT's [infrastructure](https://github.com/llvm/circt/blob/main/include/circt/Scheduling/Problems.h) for modeling scheduling problems and algorithms? I think it would be great to (eventually) develop what is needed for HIR in the common infrastructure if it doesn't exist yet.
> 
> I saw a mention of `createStandardToHIRPass` but I didn't see its implementation. In general, I am curious what dialects you are interested in lowering to HIR. I am currently interested in using the scheduling infrastructure on affine loops, and then lowering these to this [pipeline while op](https://circt.llvm.org/docs/Dialects/StaticLogic/#staticlogicpipelinewhile-circtstaticlogicpipelinewhileop) representation, which captures dataflow but leaves exact timing to be determined by lowering to Calyx (which is in progress). Another design point we have considered on this path is using something like HIR, which I am interested in exploring. Once the dialect lands, I'd be interested in doing something like the above but targetting HIR instead of the pipeline while op.

**Comment by kingshukmajumder** [comment]
Apologies for the late reply. 

> > First, I have some logistical suggestions, since this encompasses a few things. I don't know what the policy is for wholesale additions of a dialect, but this is quite a bit to review at once. It might help to separate this into several patches along these lines:
> > 
> > * Boilerplate to create new dialect without IR, Analysis, Transforms, Conversions, or Translations
> > * IR with roundtrip lit tests through circt-opt
> > * The rest of the libraries, in separate PRs, with their respective tests

Thanks for the suggestions. This is the first time I am raising a PR so still getting used to the process :). I will break these into separate commits and PRs as you suggested.

> > Regardless, I will try to take a deeper look at this. For now I had a couple high-level questions.
> > Could the `ScheduleAnalysis` make use of CIRCT's [infrastructure](https://github.com/llvm/circt/blob/main/include/circt/Scheduling/Problems.h) for modeling scheduling problems and algorithms? I think it would be great to (eventually) develop what is needed for HIR in the common infrastructure if it doesn't exist yet.


ScheduleAnalysis is a quick check to find bugs in the IR schedule (only a subset of scheduling bugs). I will check the circt scheduling infra to see if there are anything in it that is related. My understanding is that the scheduling infra is focused on finding schedules, not verifying them.


> > I saw a mention of `createStandardToHIRPass` but I didn't see its implementation. In general, I am curious what dialects you are interested in lowering to HIR. I am currently interested in using the scheduling infrastructure on affine loops, and then lowering these to this [pipeline while op](https://circt.llvm.org/docs/Dialects/StaticLogic/#staticlogicpipelinewhile-circtstaticlogicpipelinewhileop) representation, which captures dataflow but leaves exact timing to be determined by lowering to Calyx (which is in progress). Another design point we have considered on this path is using something like HIR, which I am interested in exploring. Once the dialect lands, I'd be interested in doing something like the above but targetting HIR instead of the pipeline while op.

I removed the implementation of standardtoHIR but might have not cleaned up well. I am currently interested in lowering Affine dialect to HIR. That way HIR can reuse the memory dependence analysis work happening in Affine dialect.

**Comment by jopperm** [comment]
> > Could the `ScheduleAnalysis` make use of CIRCT's [infrastructure](https://github.com/llvm/circt/blob/main/include/circt/Scheduling/Problems.h) for modeling scheduling problems and algorithms? I think it would be great to (eventually) develop what is needed for HIR in the common infrastructure if it doesn't exist yet.
> 
> ScheduleAnalysis is a quick check to find bugs in the IR schedule (only a subset of scheduling bugs). I will check the circt scheduling infra to see if there are anything in it that is related. My understanding is that the scheduling infra is focused on finding schedules, not verifying them.

Not necessarily, the verification of a solution is an integral part in the `Problem` classes, so you could indeed construct a suitable problem for HIR, feed in the schedule you want to verify, and invoke `Problem::verify()`. Ping me on Discord if you have any questions about getting along with the scheduling infra (we don't have a high-level introduction/rationale document yet, sorry).
